reactor-core/src/main/java/reactor/core/publisher/MonoIgnoreThen.java

```diff
@@ -78,244 +73,289 @@ public Object scanUnsafe(Attr key) {
+                    }
+                    if (hasRequest(state)) {
+                        return;
+                    }
+                    if (STATE.compareAndSet(this, state, state | HAS_REQUEST)) {
+                        if (hasValue(state)) {
+                            final CoreSubscriber<? super T> actual = this.actual;
+                            final T v = this.value;
+
+                            actual.onNext(v);
+                            actual.onComplete();
+                        }
+                        return;
+                    }
+                }
+            }
         }
-        
+
         @Override
-        public void onError(Throwable t) {
-            this.parent.onError(t);
+        public void onNext(T t) {
+            if (this.done) {
+                Operators.onDiscard(t, currentContext());
+                return;
+            }
+
+            if (this.index != this.ignoreMonos.length) {
+                // ignored
+                Operators.onDiscard(t, currentContext());
+                return;
+            }
+
+            this.done = true;
+
+            complete(t);
         }
-        
+
         @Override
         public void onComplete() {
-            this.parent.ignoreDone();
-        }
-        
-        void cancel() {
-            Operators.terminate(S, this);
-        }
-        
-        void clear() {
-            S.lazySet(this, null);
-        }
-    }
-    
-    static final class ThenAcceptInner<T> implements InnerConsumer<T> {
-        final ThenIgnoreMain<T> parent;
-        
-        volatile Subscription s;
-        @SuppressWarnings("rawtypes")
-        static final AtomicReferenceFieldUpdater<ThenAcceptInner, Subscription> S =
-                AtomicReferenceFieldUpdater.newUpdater(ThenAcceptInner.class, Subscription.class, "s");
+            if (this.done) {
+                return;
+            }
 
-        boolean done;
-        
-        ThenAcceptInner(ThenIgnoreMain<T> parent) {
-            this.parent = parent;
+            if (this.index != this.ignoreMonos.length) {
+                final int previousState = markUnsubscribed();
+                if (isCancelled(previousState)) {
+                    return;
+                }
+                this.activeSubscription = null;
+                this.index++;
+                subscribeNext();
+                return;
+            }
+
+            this.done = true;
+
+            this.actual.onComplete();
         }
 
-        @Override
-        @Nullable
-        public Object scanUnsafe(Attr key) {
-            if (key == Attr.PARENT) return s;
-            if (key == Attr.ACTUAL) return parent;
-            if (key == Attr.TERMINATED) return done;
-            if (key == Attr.CANCELLED) return s == Operators.cancelledSubscription();
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        void subscribeNext() {
+            final Publisher<?>[] a = this.ignoreMonos;
+
+            for (;;) {
```

Comment:
I was trying to see if another loop construct was possible closer to the part where we evaluate if `this.ignoreMonos[]` are callable, but there will always be a need to do one last loop if the last in the array is `Callable` (since we'd then need to evaluate `this.lastMono`, so let's keep this method that way 👍